src/families/tron/bridge/js.js

```diff
@@ -259,20 +240,90 @@ const currencyBridge: CurrencyBridge = {
   scanAccountsOnDevice
 };
 
-const createTransaction = a => {
-  throw new CurrencyNotSupported("tron currency not supported", {
-    currencyName: a.currency.name
-  });
-};
+const createTransaction = () => ({
+  family: "tron",
+  amount: BigNumber(0),
+  mode: "send",
+  duration: 3,
+  recipient: "",
+  networkInfo: null
+});
 
 const updateTransaction = (t, patch) => ({ ...t, ...patch });
 
-const getTransactionStatus = a =>
-  Promise.reject(
-    new CurrencyNotSupported("tron currency not supported", {
+const getEstimatedFees = async t => {
+  // see : https://developers.tron.network/docs/bandwith#section-bandwidth-points-consumption
+  // 1. cost around 200 Bandwidth, if not enough check Free Bandwidth
+  // 2. If not enough, will cost some TRX
+  // 3. normal transfert cost around 0.002 TRX
+  // Special case: If activated an account, cost around 0.1 TRX
+  let fees = BigNumber(0);
+
+  // Calculate Bandwidth :
+  if (t.networkInfo) {
+    const {
+      freeNetUsed = 0,
+      freeNetLimit = 0,
+      NetUsed = 0,
+      NetLimit = 0
+    } = t.networkInfo;
+    const bandwidth = freeNetLimit - freeNetUsed + NetLimit - NetUsed;
+
+    if (bandwidth < AVERAGE_BANDWIDTH_COST) {
+      fees = fees.plus(BigNumber(2000));
+      // Cost is around 0.002 TRX
+    }
+  }
+
+  // Active the account if there's no data, that require fees
+  if (t.recipient) {
+    const recipientAccount = await fetchTronAccount(t.recipient);
+    if (recipientAccount.length === 0) {
+      fees = fees.plus(BigNumber(100000));
+      // Cost is around 0.1 TRX
+    }
+  }
+
+  return fees;
+};
+
+const getTransactionStatus = async (a, t) => {
+  const errors = {};
+  const warnings = {};
+  const tokenAccount = !t.subAccountId
+    ? null
+    : a.subAccounts && a.subAccounts.find(ta => ta.id === t.subAccountId);
+  const account = tokenAccount || a;
+
+  if (!t.recipient) {
+    errors.recipient = new RecipientRequired("");
```

Comment:
To avoid to mutate errors, i suggest to have private functions which handle with each kind of errors like `getRecipientError: (recipient: string) => Promise<?Error>`.

After that, you could construct your `errors: { [string]: Error }` map in a more typesafe way

Reply:
Most of the time I would argue that mutation style is not so great.
Here I think it's ok to mutate from inside the validation logic (obviously not outside), maybe explicit typing would help the type missing for sure.

I can see the benefit of going mutation when we have lot of tricky logic (e.g. https://github.com/LedgerHQ/ledger-live-common/blob/master/src/families/tezos/bridge/libcore.js#L180-L205 )

one interesting fact is that you look at the function and the whole validation is at this same place for all form fields validations. For now that's what we have but i could see the advantage of splitting pieces if we see a pattern of code sharing more the logic in the future. 🤷‍♂️

Reply:
I see the point yes, in these case, i agree that it's an OK tradeoff.